From b0e57012bfcb5caf656cdec38b2abc8a91839f88 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Thu, 29 Jun 2023 15:14:37 +0200
Subject: [PATCH 0486/1695] clk: stm32mp25: put all fexgen clocks as critical

Actually in A35 TDCID mode, all flexgen are managed by A35, M33 does
not use an SCMI agent to enable the flexgen yet. So we have to enable
all of them for the M33.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 99830f0f1c789d31e9cfa5a7d405ea5fee13faca clk: stm32mp25: put all fexgen clocks as critical]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I2ac9a950e75131a8b135baa50eead107ded66a2f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/326483
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 core/drivers/clk/clk-stm32mp25.c | 28 +++++++++++++++++++++-------
 1 file changed, 21 insertions(+), 7 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index ce57f984f..9bc875f63 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -3996,6 +3996,24 @@ struct clk *stm32mp25_clk_provided[STM32MP25_ALL_CLK_NB] = {
 	[TXBYTECLK]		= &txbyteclk,
 };
 
+static void clk_stm32_set_flexgen_as_critical(void)
+{
+	uint32_t i = 0;
+
+	for (i = 0; i < XBAR_CHANNEL_NB; i++) {
+		unsigned long clock_id = CK_ICN_HS_MCU + i;
+		struct clk *clk;
+
+		if (!stm32_rcc_has_access_by_id(i))
+			continue;
+
+		clk = stm32mp_rcc_clock_id_to_clk(clock_id);
+		assert(clk);
+
+		clk_enable(clk);
+	}
+}
+
 static bool clk_stm32_clock_is_critical(__maybe_unused struct clk *clk)
 {
 	struct clk *clk_criticals[] = {
@@ -4006,13 +4024,6 @@ static bool clk_stm32_clock_is_critical(__maybe_unused struct clk *clk)
 		&ck_lse,
 		&ck_cpu1,
 		&ck_icn_p_syscpu1,
-		&ck_icn_hs_mcu,
-		&ck_icn_ls_mcu,
-		&ck_icn_sdmmc,
-		&ck_icn_ddr,
-		&ck_icn_hsl,
-		&ck_icn_nic,
-		&ck_icn_vid,
 		&ck_icn_s_ddr,
 		&ck_icn_p_ddrc,
 		&ck_icn_p_ddrcfg,
@@ -4476,6 +4487,9 @@ end:
 		}
 	}
 
+	/* TEMPORARY: Enables all root clocks (could be used by M33) */
+	clk_stm32_set_flexgen_as_critical();
+
 	return TEE_SUCCESS;
 }
 
-- 
2.47.2

